### PR TITLE
[CR] [WIP] vehicle: more iterations for part collision loop (hotfix for APC-vs-shrubbery)

### LINKED ISSUE #19571: Greenery a little too durable?
I've been driving around a 19-ton APC (yes, you read that correctly) with a somewhat underpowered engine (V6 gasoline) that can hit around 32 km/h or 20 mph. Now, it consistently fails to ram through more than one shrub at a time due to speed loss which I find very unlikely... but in theory possible under less-than-ideal circumstances.

However, this same 19-ton vehicle managed to be brought to a complete stop from a speed of 22 km/h (~14 mph) by, of all things, a datura flower. Not even a sapling, but a **flower**. A flower stopped a 19-ton vehicle.

I'm starting to think there's something broken when it comes to the game's ability to account for mass in collisions.

### LINKED ISSUE #19854: Vehicle mass is apparently ignored
This really needs looking into with the new lower speed on vehicles. As it stands speed is the only thing that counts to the point that the easiest way to get my 4 or 5 ton 40 mph vehicle through the brush is to go out and smash bushes to clear the way. And I'm struggling to squish even regular zombies,  which is just wrong. Where as a light electric vehicle going 70 mph or so pulps the hell out of brutes...and kinda wrecks itself in the process. Which means that you can't really build a zombie killing death machine anymore. And considering the effort, time, and materials involved is a distinct disappointment.

Don't get me wrong, I'm all for the vehicle rebalance, just it's exposing some flaws in the system.

### PR DESCRIPTION
**Pull request by keyspace**
Closes #19854.
Closes #19571.

### Summary

There are two factors that determine how fast speed drops upon collision:

* momentum (scalar magnitude), calculation [in this block](https://github.com/CleverRaven/Cataclysm-DDA/blob/32ebedc26456275e98c6c2db5621dea6d58c5882/src/vehicle.cpp#L4434);
* general scale-down, seemingly to guarantee a loop exit [here](https://github.com/CleverRaven/Cataclysm-DDA/blob/32ebedc26456275e98c6c2db5621dea6d58c5882/src/vehicle.cpp#L4553).

This PR adjusts the latter.

### Current implementation details (none of my doing)

Debug notes got scrubbed, so re-creating from hazy memory a week later, and re-doing research.

* Collisions in general are detected in `map.cpp`, ` map::move_vehicle()`. There's a `do-while` loop, iterated 10 (magic!) times, that calls `vehicle::collision()`. This loop also seems to apply damage; the function in general fiddles with the `facing` direction, and many more things.
* `vehicle::collision()` iterates over all part locations and calls `vehicle::part_collision()` on them. This happens whether there was an "actual collision" or not. I.e. moving over "flat terrain" with nothing on it still performs the following call. The main purpose of these two functions is _checking_.
* `vehicle::part_collision()` detects if a specific vehicle tile has collided, if so - which part, determines the collision type first (squishy body, impassable terrain, etc.). If there is no collision, it [returns early](https://github.com/CleverRaven/Cataclysm-DDA/blob/32ebedc26456275e98c6c2db5621dea6d58c5882/src/vehicle.cpp#L4386), which is then propagated "up" and can ["return" early, too](https://github.com/CleverRaven/Cataclysm-DDA/blob/32ebedc26456275e98c6c2db5621dea6d58c5882/src/vehicle.cpp#L4228), and so on. These early returns are the only things that keep the functions at bay performance-wise.
  * If there _was_ a collision detected in `vehicle::part_collision()`, the function proceeds to a `do-while` loop that _calculates_ damage to vehicle parts and the collided object. On each iteration, the loop reduces the vehicle's speed. It colliding object was a body (creature), this loop also _applies_ damage to it (but not the vehicle!), stuns it, flings it around, etc.. In either case, this loop continues until [either the vehicle parts or colliding object are smashed, or vehicle speed drops to zero](https://github.com/CleverRaven/Cataclysm-DDA/blob/32ebedc26456275e98c6c2db5621dea6d58c5882/src/vehicle.cpp#L4555).
  * Note: although code reads "velocity" (vector) in this section, it is actually speed (scalar). References to "impulse" (force times time) are also misleading - although [dimensionally equivalent]( https://en.wikipedia.org/wiki/Momentum), it's momentum (mass times velocity) - or, to be precise, momentum magnitude.
  * The `do-while` previously had a "divide by 100 at start, multiply by 100 at end" construct, with and [optional](https://github.com/CleverRaven/Cataclysm-DDA/blob/32ebedc26456275e98c6c2db5621dea6d58c5882/src/vehicle.cpp#L4449) "multiply by 90" instead if the two-body system managed to _gain_ energy in a given iteration. The actual origin of this down-up scaling is lost git history.
    * In #15804 the "multiply by 100 at end" part got replaced with "multiply by 90", effectively introducing a [geometric progression](https://en.wikipedia.org/wiki/Geometric_progression) with a common ratio of 0.9. From then on, **there has been a second influence on the speed drop during a collision**.
    * For heavy vehicles, momentum magnitude delta (and resulting speed delta) upon collision is quite negligible. However, due to above geometric progression an infinite-mass vehicle would still stop, and pretty fast. Here's a [basic script](https://gist.github.com/keyspace/085c74e20786bb128a9e9345c6661498) to demonstrate the effect of ratio in question in such a case.

### Script output (infinite-mass vehicle!)

First line - speeds (in arbitrary units), second line - loop iterations until speed drops to `floor(0)`.

r = 0.8, proposed for "unexpected energy increase":

```
[10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200]
[11, 14, 16, 17, 18, 19, 20, 20, 21, 21, 22, 22, 22, 23, 23, 23, 24, 24, 24, 24]
```

r = 0.9, current for both "unexpected energy increase" and "unconditional scaling":

```
[10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200]
[22, 29, 33, 36, 38, 39, 41, 42, 43, 44, 45, 46, 47, 47, 48, 49, 49, 50, 50, 51]
```

r = 0.95:

```
[10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200]
[45, 59, 67, 72, 77, 80, 83, 86, 88, 90, 92, 94, 95, 97, 98, 99, 101, 102, 103, 104]
```

r = 0.97, proposed for "unconditional scaling":

```
[10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200]
[76, 99, 112, 122, 129, 135, 140, 144, 148, 152, 155, 158, 160, 163, 165, 167, 169, 171, 173, 174]
```

r = 0.99, originally proposed for both "unexpected energy increase" and "unconditional scaling":

```
[10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200]
[230, 299, 339, 368, 390, 408, 423, 437, 448, 459, 468, 477, 485, 492, 499, 505, 512, 517, 523, 528]
```

-----

EDIT42: completely rewritten
EDIT44: fixed links
EDIT45: preparation for post-mortem

### CODE REVIEW

**Comment by keyspace** [comment]
Tested sloppily: riding an Electric SUV into the forest stops dead at the second tree. Doing the same with an APC makes a tunnel through the overmap tile in about 3 full-speed rams.

File: src/vehicle.cpp [hunk 1]
```diff
@@ -4273,7 +4275,7 @@ veh_collision vehicle::part_collision( int part, const tripoint &p,
             }
         }
 
-        coll_velocity = vel1_a * ( smashed ? 100 : 90 );
+        coll_velocity = vel1_a * ( smashed ? 100 : 99 );
```
Comment (keyspace) [inline src/vehicle.cpp]: Can be improved further by not scaling here. As mentioned in PR description, I don't know (yet) why it's like this in the first place, so leaving.

**Comment by keyspace** [comment]
Intentionally as minimal as possible, so marking [RDY].

**Comment by Coolthulhu** [comment]
Last time I checked, the performance loop was called quite a lot.
This will need more testing than just if "it works":
* How does it behave at high velocity
* How does it behave at low velocity
* How does it behave at high velocity against a very resistant target (say, a hulk with 1000% hp world setting)

**Comment by keyspace** [comment]
`vehicle::part_collision()` (where the loop resides) is called each turn, but it exits early if there was no collision (`veh_coll_nothing`). Are you referring to that, or this exact `do {...} while (...)` loop? The latter has mostly floating point math, and several terrain/critter/etc lookups.

I'll try a profiler on this section when I have time, but that didn't work well for me last time I tried it.

-----

Tangent: loop also has check for dislocated monster:

``` c++
smashed = (critter->pos() != p);
```

Remember this mentioned in an issue about monsters ending up inside vehicle on impact.

**Comment by keyspace** [comment]
f97c0779925908b7d0016e1fbc103c65e779d7b2 (part of  #15804) was the commit that introduced the geometric progression, and the particular ratio of 0.9.

File: src/vehicle.cpp (hunk 1 above)
Comment (keyspace) [inline src/vehicle.cpp]: Actually, see #19974 discussion - it was added for a reason. "Improved" is loaded.

**Comment by keyspace** [comment]
Rebased the branch on top of a post-VPO-revert commit (that I use for "live testing"), and completely rewritten the preamble. It may serve as a general introduction to collisions we have now - for anyone trying to understand this PR, or re-implement collisions in the future.

Requested testing with different vehicles/scenarios still pending.

**Comment by keyspace** [comment]
I've also pushed a new commit that reduces the proposed common ratio from 0.99 to 0.97. Although I've much liked 0.99 in my "live tests", and collisions in general haven't exceeded 40-70 (when this ratio was used!) for car-brand-car at 96 km/h, electric SUV at 64 km/h and APC at 16 km/h, I feel that the numbers may seem scary.

Haven't done testing with 0.97 yet, will get back when I do.

**Comment by keyspace** [comment]
Had links in preamble all screwed up - hopefully fixed.

-----

Did some in-game testing for behaviour at infra-slow speeds. Save attached. [test-vehicles.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/706986/test-vehicles.zip)

Tangent: good candidate for automated testing.

General idea: can a vehicle overcome an obstacle if speeding up from 0 to 16 km/h, starting point-blank from the obstacle?

Note: 16 km/h ~ 10 mph, one-forward or one-reverse.

Tried it in tandem with my [WIP fix for vehicles-vs-flowers](https://github.com/keyspace/Cataclysm-DDA/pull/1) (see #20017 for notes) to save time. Tell me if I screwed up badly.

`r` - common ratio changed in this PR, [this line](https://github.com/CleverRaven/Cataclysm-DDA/blob/32ebedc26456275e98c6c2db5621dea6d58c5882/src/vehicle.cpp#L4555).

`^` and `v` - forward or backward.

`f` and `#` - dandelions and shrubs.

`y` and `n` - whether vehicle crushes it.

Numbers are `do-while` loop iteration counts. They're not consistent! I've had them different for the same vehicle, same `r`, on different tries. The yes-no results were consistent, though.

Note: [empty table](https://gist.githubusercontent.com/keyspace/913daa7bced869391ef85dca4c7661ab/raw/de485d15452991926552c30c5e9e06c39ae7086e/table.md).

#### r = 90 (current git master):

vehicle              | ^f    | vf    | ^#    | v#
:------------------- | :---: | :---: | :---: | :---:
car                  | n/a   | n/a   | 11 y  | 39 n
electric_car         | n/a   | n/a   | 40 n  | 38 n
bicycle              | n/a   | n/a   | 15 n  | 14 n
motorcycle           | n/a   | n/a   | 26 n  | 24 n
military_cargo_truck | n/a   | n/a   | 7 y   | 10 y


#### r = 97 (in current PR):

vehicle              | ^f    | vf    | ^#    | v#
:------------------- | :---: | :---: | :---: | :---:
car                  | 10 y  | 104 n | 12 y  | 31 y
electric_car         | 28 y  | 99 n  | 13 y  | 74 n
bicycle              | 44 n  | 40 n  | 19 n  | 18 n
motorcycle           | 72 n  | 64 n  | 38 n  | 35 n
military_cargo_truck | 12 y  | 10 y  | 9 y   | 11 y


#### r = 98:

vehicle              | ^f    | vf    | ^#    | v#
:------------------- | :---: | :---: | :---: | :---:
car                  | 12 y  | 29 y  | 7 y   | 16 y
electric_car         | 19 y  | 126 n | 22 y  | 53 y
bicycle              | 48 n  | 44 n  | 20 n  | 18 n
motorcycle           | 84 n  | 74 n  | 40 n  | 37 n
military_cargo_truck | 10 y  | 7 y   | 8 y   | 5 y


#### r = 99 (initially in PR):

vehicle              | ^f    | vf    | ^#    | v#
:------------------- | :---: | :---: | :---: | :---:
car                  | 7 y   | 25 y  | 8 y   | 20 y
electric_car         | 16 y  | 41 y  | 12 y  | 33 y
bicycle              | 54 n  | 49 n  | 21 n  | 19 n
motorcycle           | 103 n | 90 n  | 44 n  | 40 n
military_cargo_truck | 10 y  | 10 y  | 6 y   | 12 y

Thoughts on results below.

**Comment by keyspace** [comment]
P.S. Test game setup:

test-vehicles

Includes electric scooter - didn't know how borked it was. Ignore it.

Military cargo truck has dandelions _under_ it due to wheel configuration. Only relevant for "vehicle-vs-flowers", no influence otherwise.

MCT chosen over APC due to non-obstructed view from cabin, similar safe speed and similar mass.

**Comment by keyspace** [comment]
1. The VPO revert has fixed heavy vehicles getting unexpected blockage from shrubs. As far as that goes, the issues referenced in preamble can be closed.
2. The underlying issue remains: there is a (reverse) geometric progression that affects slow-moving vehicles. `car` can't overcome shrub in reverse. `electric_car` can neither in forward nor reverse.
3. 0.97 common ratio is sufficient not to block `car` on shrub in reverse, but insufficient to do the same for `electric_car`. Point of contention - should it?..
4. `bicycle` and `motorcycle` can't overcome shrubs in _any_ case.
5. Everything gets stuck on dandelions more than shrubs. Possibly caused by my WIP "vehicle-vs-flowers" re-setting part to collide - [wheels instead of frame](https://github.com/keyspace/Cataclysm-DDA/blob/2938ba48c8c7834f1098b0e76914f4d383ed4b5d/src/vehicle.cpp#L4370). Not relevant to this PR.

-----

6. Performing this test took me less time than it would've learning Catch. I will, I promise.

File: src/vehicle.cpp [hunk 2]
```diff
@@ -4440,13 +4441,15 @@ veh_collision vehicle::part_collision( int part, const tripoint &p,
         const float E_before = 0.5f * (mass * vel1 * vel1)   + 0.5f * (mass2 * vel2 * vel2);
         const float E_after =  0.5f * (mass * vel1_a*vel1_a) + 0.5f * (mass2 * vel2_a*vel2_a);
         const float d_E = E_before - E_after;
+
         if( d_E <= 0 ) {
             // Deformation energy is signed
             // If it's negative, it means something went wrong
             // But it still does happen sometimes...
+            // @todo: debugmsg?..
```
Comment (Coolthulhu) [inline src/vehicle.cpp]: Or "soft debugmsg": `add_msg( m_debug, ... )`

**Comment by Coolthulhu** [comment]
I see tests with furniture (flowers), but not with creatures. Creatures are also important - they should be (preferably) pushed away, but shouldn't be flattened by too light vehicles.

The code for pushing away already exists, but the problem with it is that it doesn't accumulate as well as damage. This may require a custom solution, just for the creatures.

A regular car may flatten zombies by pushing against them, but if same thing happens to hulks, it's not OK.

**Comment by keyspace** [comment]
> I see tests with furniture (flowers), but not with creatures.

Yup, that's just one test of the three you mentioned earlier. Will do all, eventually.

> A regular car may flatten zombies by pushing against them, but if same thing happens to hulks, it's not OK.

Understood. Will try with creatures from different `MS_<SIZE>` groups, I guess.

**Comment by keyspace** [comment]
> I see tests with furniture (flowers), but not with creatures.

Re-read. Slow speeds, push away creatures. Got it.

**Comment by keyspace** [comment]
This is no longer a hotfix. As noted [some comments above](https://github.com/CleverRaven/Cataclysm-DDA/pull/19974#issuecomment-272749881), the bugs in preamble, although still _there_, are no longer dire after the VPO revert.

This is instead a crusade to change a magic number. It's also stalled until I (or someone else) finds the time to write automated tests.

I'm inclined to close this, and open a provisional issue about vehicle collisions (linking here for the gory details) instead.

OK?

**Comment by Coolthulhu** [comment]
I'm not sure a magic number is enough.
Vehicle collision is pretty broken because it is a hard problem. Every change needs to be tested against multiple cases and sub-cases.

**Comment by keyspace** [comment]
Yes.

I would note in the issue that collision tests are desired before collision rewrites.

**Comment by keyspace** [comment]
Closing as described above. Will save Jenkins time, too.

**Pull request closed without merging**